### PR TITLE
Remove cached node bytes from merkle nodes 

### PR DESCRIPTION
**Pull request by dboehm-avalabs**
We mostly load nodes just before we edit them, so holding on to the bytes just wastes memory.

### CODE REVIEW

File: x/merkledb/codec.go [hunk 1]
```diff
@@ -91,16 +89,41 @@ type codecImpl struct {
...
 
+func childSize(index byte, childEntry *child) int {
+	return uintSize(uint64(index)) + len(ids.Empty) + keySize(childEntry.compressedKey) + boolSize()
+}
+
+func boolSize() int {
+	return 1
+}
+
+var log128 = math.Log(128)
```
Comment (danlaine) [inline x/merkledb/codec.go:110]: We typically put all package level vars in a `var` block at the top of the file

File: x/merkledb/codec.go [hunk 2]
```diff
@@ -91,16 +89,41 @@ type codecImpl struct {
...
+	total := uintSize(uint64(len(n.children))) + boolSize() + len(n.value.Value())
+	// for each non-nil entry, we add the additional size of the child entry
+	for index, entry := range n.children {
+		total += childSize(index, entry)
+	}
+	return total
+}
 
+func childSize(index byte, childEntry *child) int {
+	return uintSize(uint64(index)) + len(ids.Empty) + keySize(childEntry.compressedKey) + boolSize()
```
Comment (danlaine) [inline x/merkledb/codec.go:103]:
```suggestion
	return uintSize(uint64(index)) +ids.IDLen + keySize(childEntry.compressedKey) + boolSize()
```

**Review by danlaine — COMMENTED** [review]
Is there a benchmark that measures the effect of this change?

**Comment by dboehm-avalabs** [comment]
> Is there a benchmark that measures the effect of this change?

Not explicitly.  This is mostly to reduce the size of cache entries since they currently store the node's bytes twice, once as the raw bytes and once in the various fields on the node struct.  Removing this extra copy with roughly halve the memory footprint per node, so we should be able to store almost twice the nodes in the cache.

File: x/merkledb/codec.go [hunk 3]
```diff
@@ -91,16 +91,42 @@ type codecImpl struct {
...
+
+// based on the current implementation of codecImpl.encodeUint which uses binary.PutUvarint
+func (*codecImpl) uintSize(value uint64) int {
+	// catch the call to log(0) and return early
+	if value == 0 {
+		return 1
+	}
+	// binary.PutUvarint repeatedly divides by 128 until the value is under 128,
+	// so count the number of times that occurs
+	return 1 + int(math.Log(float64(value))/log128)
```
Comment (darioush) [inline x/merkledb/codec.go:119]: should we avoid floating point math here?

Comment (dboehm-avalabs) [reply x/merkledb/codec.go:119]: it isn't the end of the world if it is off, but I could switch this to a looped version

Comment (dboehm-avalabs) [reply x/merkledb/codec.go]: Done

**Review by darioush — APPROVED** [review]

File: x/merkledb/node.go [hunk 4]
```diff
@@ -66,17 +64,7 @@ func (n *node) hasValue() bool {
 
 // Returns the byte representation of this node.
 func (n *node) bytes() []byte {
-	if n.nodeBytes == nil {
```
Comment (danlaine) [inline x/merkledb/node.go:69]: Removing this cached value means we'll have to reserialize the node before writing it. Are we concerned about the additional time that'll take? Seems to be a tradeoff between memory usage and CPU usage.

Comment (dboehm-avalabs) [reply x/merkledb/node.go:69]: anything we are writing is something we edited so this would have always been nil

**Review by danlaine — APPROVED** [review]

**Pull request merged**